lib/galaxy/model/__init__.py

```diff
@@ -6499,6 +6499,14 @@ def attribute_columns(column_collection, attributes, nesting_level=None):
         q = q.order_by(*order_by_columns)
         return q
 
+    @property
+    def elements_deleted(self):
+        stmt = self._build_nested_collection_attributes_stmt(
+            hda_attributes=("deleted",), dataset_attributes=("deleted",)
+        )
+        stmt = exists(stmt).where(or_(HistoryDatasetAssociation.deleted == true(), Dataset.deleted == true()))
```

Comment:
```suggestion
        stmt = exists(stmt.subquery()).where(or_(HistoryDatasetAssociation.deleted == true(), Dataset.deleted == true()))
```

Reply:
That's not what the docs do: https://docs.sqlalchemy.org/en/20/tutorial/data_select.html#exists-subqueries and it also blows up with:
```
File ~/src/galaxy/lib/galaxy/model/__init__.py:6507, in DatasetCollection.elements_deleted(self)
   6502 @property
   6503 def elements_deleted(self):
   6504     stmt = self._build_nested_collection_attributes_stmt(
   6505         hda_attributes=("deleted",), dataset_attributes=("deleted",)
   6506     )
-> 6507     stmt = exists(stmt.subquery()).where(or_(HistoryDatasetAssociation.deleted == true(), Dataset.deleted == true()))
   6508     return object_session(self).execute(select(stmt)).scalar()

File ~/src/galaxy/.venv/lib/python3.10/site-packages/sqlalchemy/sql/_selectable_constructors.py:181, in exists(__argument)
    139 def exists(
    140     __argument: Optional[
    141         Union[_ColumnsClauseArgument[Any], SelectBase, ScalarSelect[Any]]
    142     ] = None,
    143 ) -> Exists:
    144     """Construct a new :class:`_expression.Exists` construct.
    145
    146     The :func:`_sql.exists` can be invoked by itself to produce an
   (...)
    178
    179     """  # noqa: E501
--> 181     return Exists(__argument)

File ~/src/galaxy/.venv/lib/python3.10/site-packages/sqlalchemy/sql/selectable.py:6646, in Exists.__init__(self, _Exists__argument)
   6644     s = __argument
   6645 else:
-> 6646     s = Select(__argument).scalar_subquery()
   6648 UnaryExpression.__init__(
   6649     self,
   6650     s,
   (...)
   6653     wraps_column_expression=True,
   6654 )

File ~/src/galaxy/.venv/lib/python3.10/site-packages/sqlalchemy/sql/selectable.py:3558, in SelectBase.scalar_subquery(self)
   3555 if self._label_style is not LABEL_STYLE_NONE:
   3556     self = self.set_label_style(LABEL_STYLE_NONE)
-> 3558 return ScalarSelect(self)

File ~/src/galaxy/.venv/lib/python3.10/site-packages/sqlalchemy/sql/selectable.py:6502, in ScalarSelect.__init__(self, element)
   6500 def __init__(self, element: SelectBase) -> None:
   6501     self.element = element
-> 6502     self.type = element._scalar_type()
   6503     self._propagate_attrs = element._propagate_attrs

File ~/src/galaxy/.venv/lib/python3.10/site-packages/sqlalchemy/sql/selectable.py:5161, in Select._scalar_type(self)
   5159     return NULLTYPE
   5160 elem = self._raw_columns[0]
-> 5161 cols = list(elem._select_iterable)
   5162 return cols[0].type

File ~/src/galaxy/.venv/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py:1130, in _non_memoized_property.__get__(self, obj, cls)
   1128 if obj is None:
   1129     return self
-> 1130 return self.fget(obj)

File ~/src/galaxy/.venv/lib/python3.10/site-packages/sqlalchemy/sql/selectable.py:930, in FromClause._select_iterable(self)
    928 @util.ro_non_memoized_property
    929 def _select_iterable(self) -> _SelectIterable:
--> 930     return (c for c in self.c if not _never_select_column(c))

File ~/src/galaxy/.venv/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py:1141, in _memoized_property.__get__(self, obj, cls)
   1139 if obj is None:
   1140     return self
-> 1141 obj.__dict__[self.__name__] = result = self.fget(obj)
   1142 return result

File ~/src/galaxy/.venv/lib/python3.10/site-packages/sqlalchemy/sql/selectable.py:860, in FromClause.c(self)
    858 if "_columns" not in self.__dict__:
    859     self._init_collections()
--> 860     self._populate_column_collection()
    861 return self._columns.as_readonly()

File ~/src/galaxy/.venv/lib/python3.10/site-packages/sqlalchemy/sql/selectable.py:1633, in AliasedReturnsRows._populate_column_collection(self)
   1632 def _populate_column_collection(self) -> None:
-> 1633     self.element._generate_fromclause_column_proxies(self)

File ~/src/galaxy/.venv/lib/python3.10/site-packages/sqlalchemy/sql/selectable.py:6316, in Select._generate_fromclause_column_proxies(self, subquery, proxy_compound_columns)
   6290     prox = [
   6291         c._make_proxy(
   6292             subquery,
   (...)
   6313         if is_column_element(c)
   6314     ]
   6315 else:
-> 6316     prox = [
   6317         c._make_proxy(
   6318             subquery,
   6319             key=proxy_key,
   6320             name=required_label_name,
   6321             name_is_truncatable=True,
   6322         )
   6323         for (
   6324             required_label_name,
   6325             proxy_key,
   6326             fallback_label_name,
   6327             c,
   6328             repeated,
   6329         ) in (self._generate_columns_plus_names(False))
   6330         if is_column_element(c)
   6331     ]
   6333 subquery._columns._populate_separate_keys(prox)

File ~/src/galaxy/.venv/lib/python3.10/site-packages/sqlalchemy/sql/selectable.py:6317, in <listcomp>(.0)
   6290     prox = [
   6291         c._make_proxy(
   6292             subquery,
   (...)
   6313         if is_column_element(c)
   6314     ]
   6315 else:
   6316     prox = [
-> 6317         c._make_proxy(
   6318             subquery,
   6319             key=proxy_key,
   6320             name=required_label_name,
   6321             name_is_truncatable=True,
   6322         )
   6323         for (
   6324             required_label_name,
   6325             proxy_key,
   6326             fallback_label_name,
   6327             c,
   6328             repeated,
   6329         ) in (self._generate_columns_plus_names(False))
   6330         if is_column_element(c)
   6331     ]
   6333 subquery._columns._populate_separate_keys(prox)

File ~/src/galaxy/.venv/lib/python3.10/site-packages/sqlalchemy/sql/elements.py:4761, in Label._make_proxy(self, selectable, name, compound_select_cols, **kw)
   4753 # there was a note here to remove this assertion, which was here
   4754 # to determine if we later could support a use case where
   4755 # the key and name of a label are separate.  But I don't know what
   (...)
   4758 # is attempting to disambiguate an explicit label, which is not what
   4759 # the user would want.   See issue #6090.
   4760 if key != self.name and not isinstance(self.name, _anonymous_label):
-> 4761     raise exc.InvalidRequestError(
   4762         "Label name %s is being renamed to an anonymous label due "
   4763         "to disambiguation "
   4764         "which is not supported right now.  Please use unique names "
   4765         "for explicit labels." % (self.name)
   4766     )
   4768 e._propagate_attrs = selectable._propagate_attrs
   4769 e._proxies.append(self)

InvalidRequestError: Label name deleted is being renamed to an anonymous label due to disambiguation which is not supported right now.  Please use unique names for explicit labels.
```

If you want I can do the alternative form of:
```diff
diff --git a/lib/galaxy/model/__init__.py b/lib/galaxy/model/__init__.py
index cbf60fcc93..25ab2dc043 100644
--- a/lib/galaxy/model/__init__.py
+++ b/lib/galaxy/model/__init__.py
@@ -6504,7 +6504,7 @@ class DatasetCollection(Base, Dictifiable, UsesAnnotations, Serializable):
         stmt = self._build_nested_collection_attributes_stmt(
             hda_attributes=("deleted",), dataset_attributes=("deleted",)
         )
-        stmt = exists(stmt).where(or_(HistoryDatasetAssociation.deleted == true(), Dataset.deleted == true()))
+        stmt = stmt.exists().where(or_(HistoryDatasetAssociation.deleted == true(), Dataset.deleted == true()))
         return object_session(self).execute(select(stmt)).scalar()
```
but both generate the same query (without warnings that you may have in mind from the old migration ?):
```sql
SELECT EXISTS (SELECT history_dataset_association.deleted AS deleted, dataset.deleted AS deleted
FROM dataset_collection AS dataset_collection_1 JOIN dataset_collection_element AS dataset_collection_element_1 ON dataset_collection_element_1.dataset_collection_id = dataset_collection_1.id JOIN history_dataset_association ON history_dataset_association.id = dataset_collection_element_1.hda_id JOIN dataset ON dataset.id = history_dataset_association.dataset_id
WHERE dataset_collection_1.id = %(id_1)s AND (history_dataset_association.deleted = true OR dataset.deleted = true) ORDER BY dataset_collection_element_1.element_index) AS anon_1
```

Reply:
Yep, all is correct. I think I mistook it for `stmt = select(exists(stmt))` where stmt should be a subquery. Sorry!